File: src/schema/part.ts
import type { PartId, SlotTag, TextureId } from './ids'
import type { Transform } from './transform'

export type TextureChannel = 'diffuse' | 'normal' | 'specular' | 'emissive'

export type TextureSlot = {
  readonly channel: TextureChannel
  readonly defaultTextureId: TextureId | undefined
}

export type Part = {
  readonly id: PartId
  readonly name: string
  readonly tags: readonly SlotTag[]
  readonly meshFile: string
  readonly defaultOffset: Transform
  readonly textureSlots: readonly TextureSlot[]
  readonly thumbnailFile: string | undefined
}

export const partAcceptsTag = (part: Part, tag: SlotTag): boolean =>
  part.tags.some((partTag) => partTag.value === tag.value)

File: src-tauri/src/schema/part.rs
use serde::{Deserialize, Serialize};
use ts_rs::TS;

use super::ids::{PartId, SlotTag, TextureId};
use super::vec::Transform;

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, TS)]
#[ts(export)]
pub enum TextureChannel {
    #[serde(rename = "diffuse")]
    Diffuse,
    #[serde(rename = "normal")]
    Normal,
    #[serde(rename = "specular")]
    Specular,
    #[serde(rename = "emissive")]
    Emissive,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, TS)]
#[ts(export)]
pub struct TextureSlot {
    pub channel: TextureChannel,
    #[serde(rename = "defaultTextureId")]
    pub default_texture_id: Option<TextureId>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, TS)]
#[ts(export)]
pub struct Part {
    pub id: PartId,
    pub name: String,
    pub tags: Vec<SlotTag>,
    #[serde(rename = "meshFile")]
    pub mesh_file: String,
    #[serde(rename = "defaultOffset")]
    pub default_offset: Transform,
    #[serde(rename = "textureSlots")]
    pub texture_slots: Vec<TextureSlot>,
    #[serde(rename = "thumbnailFile")]
    pub thumbnail_file: Option<String>,
}
